docs/content/guide/halo/typescript.md

```diff
@@ -57,3 +54,53 @@ const client = new Client();
 :::
 
 Before data can be manipulated, a [`space`](../echo/typescript/README.md) must be obtained.
+
+## Receiving a device join invitation
+We often want to add a device to the same [HALO identity](../halo/).
+You can accept a device join invitation by providing the invitation code and the auth code to the dxos client.
+This example demonstrates how to accept a device join invitation in a console application.
+A device join invitation can be obtained from [the shell](../halo/#shell) or generated in typescript from an existing identity.
+
+```ts{7}
+import { Client } from '@dxos/client';
+import { InvitationEncoder } from '@dxos/client-protocol';
+import * as readlineSync from 'readline-sync';
+
+/** Extracts device invitation code from a URL or string. */
+const getDeviceInvitationCode = (encoded: string): string | null => {
+  if (encoded.startsWith('http')) {
+    const searchParams = new URLSearchParams(encoded.substring(encoded.lastIndexOf('?')));
+    return searchParams.get('deviceInvitationCode') ?? null;
+  }
+
+  return encoded;
+};
+
+/** Accepts a device join invitation by prompting the user for:
+ *    - invitation code
+ *    - authentication code */
+const acceptInvitation = async (client: Client) => {
+  const encodedInvitation = readlineSync.question('Please enter an invitation: ');
+  const deviceInvitationCode = getDeviceInvitationCode(encodedInvitation);
+
+  if (!deviceInvitationCode) {
```

Comment:
we can try to simplify this snippet to be less "real" and more to the point ... e.g.: how about assuming the two values are coming from process.env somehow and are already parsed? parsing the URL might be a separate snippet in a collapsed detail section of the doc

Reply:
If you're accepting an invite from composer for example, you have to join the invitation before the shell UI presents you with the auth code.